src/peakrdl_cocotb_ralgen/testcases/rw_test.py

```diff
@@ -5,68 +5,134 @@
 logger = cocotb.log
 
 
+async def rw_test_base(RAL, key, reg, wrval, verbose, foreground_write, foreground_read,):
+    """Base function to perform read-write tests on a given register.
+
+    Params:
+        RAL: The RAL instance, required for background operations.
+        key (str): Key or identifier for the register.
+        reg (dict): Register metadata containing width and masks.
+        wrval (int): The value to write to the register.
+        verbose (bool): If True, logs the read-write results.
+        foreground_write (bool): If True, use foreground write; otherwise, use background write.
+        foreground_read (bool): If True, use foreground read; otherwise, use background read.
+
+    Raises:
+        AssertionError: If the actual value read does not match the expected value.
+    """
+    r = RAL.ifc
+    addr = reg["address"]
+    donttest = reg["donttest"]
+    wmask = reg["write_mask"]
+    rmask = reg["read_mask"]
+
+    expected = wrval & ~donttest & wmask & rmask
+
+    if foreground_write:
+        await r.write(addr, reg["width"], reg["width"], wrval)
+    else:
+        for sighash in reg["signals"]:
+            RAL.background.write(
+                sighash,
+                (wrval >> sighash["low"]) & int(
+                    "1" * (sighash["high"] - sighash["low"] + 1), 2),
+            )
+
+    if foreground_read:
+        rv = await r.read(addr, reg["width"], reg["width"])
+    else:
+        rv = 0
+        for sighash in reg["signals"]:
+            rv |= RAL.background.read(sighash) << sighash["low"]
+
+    actual = rv & ~donttest & wmask
+
+    assert (
+        actual == expected
+    ), f"{key}:: Read Write Written {wrval:x}, actual(Read) {actual:x}, Expected {expected:x}, wrMask {wmask:x}, rdmask {rmask:x}, donttest = {donttest:x}"
+
+    if verbose:
+        logger.info(
+            f"Test RW: {key} wval {wrval:x} rv {rv:x} expected {expected:x} actual {actual:x}",
```

Comment:
instead of RW take a parameter in the fn invocation and use it to distinguish between RW, w1,w0 etc.